### PR TITLE
swarm: add pull request template

### PR DESCRIPTION
**Pull request by acud**
Adds a PR template to the repo for the welfare of all our contributors and reviewers alike

### CODE REVIEW

File: pull_request_template.md
Comment (nolash) [inline pull_request_template.md]: _in_ the case of

File: pull_request_template.md
Comment (nolash) [inline pull_request_template.md]: toplevel packages right?

File: pull_request_template.md
Comment (nolash) [inline pull_request_template.md]: use "in progress" inbetween

File: pull_request_template.md
Comment (nolash) [inline pull_request_template.md]: so rebase only in the end?

File: pull_request_template.md
Comment (nolash) [inline pull_request_template.md]: :+1:

File: pull_request_template.md
Comment (nolash) [inline pull_request_template.md]: I'd like to add something like "in difference of opinion focus on the arguments." and perhaps "matters of pure taste should be the discretion of thte author"

File: pull_request_template.md
Comment (nolash) [inline pull_request_template.md]: how about "including trivial intermediate commit messages" - sometimes maybe it's not a dumb idea to have a bit of a log.

File: pull_request_template.md
Comment (acud) [reply pull_request_template.md]: can you please write the complete suggested bulletpoint?

File: pull_request_template.md
Comment (acud) [reply pull_request_template.md]: TBH in the context of big pull requests, intermediate commit messages are redundant because the underlying diff is very difficult to compare anyway. in the case of small PRs intermediate commits are trivial. i'm willing to reconcile with adding a small request to include intermediate commit message _only_ in the case that they are very important

File: pull_request_template.md
Comment (acud) [reply pull_request_template.md]: ideally yes. and ideally you shouldn't need to rebase at all. i'm willing to reconcile with adding "rebasing while the pr is under review is to be done only if requested by reviewers"

File: pull_request_template.md
Comment (acud) [reply pull_request_template.md]: TBH i find the `in progress` label redundant and never use it. if the package is not marked ready for review it means it is in progress and we should not review it. not sure why we need a label for every permutation of the universe

Comment (nolash) [reply pull_request_template.md]: If it is first "ready for review" and then needs some more work I change it to "in progress" while it's not ready for review anymore. I think this protocol makes it more clear when to re-review PRs.

File: pull_request_template.md
Comment (nolash) [reply pull_request_template.md]: > big pull requests

But there shouldn't be any, right? ;)

**Review by janos — APPROVED** [review]

File: pull_request_template.md
Comment (acud) [reply pull_request_template.md]: Yep. especially on planet nolash

**Review by nonsense — APPROVED** [review]
Overall looks good to me.

File: pull_request_template.md
Comment (zelig) [reply pull_request_template.md]: this is more complicated. if the PR review is a longer process, it is super important to keep it rebased, since it can bring up incompatibilities etc.
Another reason not to keep review process too long and PRs to be small. 
@janos you usually do a merge what do u think is the best way?

Comment (janos) [reply pull_request_template.md]: This point is fine as a warning, even I doubt that it can be always applied, but at least it would minimize force pushing. As we cannot restore old history after force pushing, I do not see what we can do when that happens. It has a bit of conflict with point number 1. With preemptive PRs, we are encouraging early comments, which means that this point would be activated very early. 

I think that rebasing and force pushing breaks history for other clones, especially when conflicts need to be resolved afterwards repeatedly. Having a merge commit allows better automatic conflict resolution by not loosing commits and history. But, I have nothing against force pushing as it provides cleaner history. It is completely ok to do force push for me, as it is a usual operation on swarm codebase. I would prefer forbidding force pushing completely, as we are squashing commits at the end, but that would be too drastic.

Comment (nolash) [reply pull_request_template.md]: Should we encourage development in parallell temporary files where possible? That way we have less issues with rebase.

Comment (acud) [reply pull_request_template.md]: My idea is to avoid force pushing as much as possible, that is why I added the reservations that the reviewer can ask to do so once the review process has started.

Rebasing brings a plethora of problems ranging from regressions due to human error while rebasing to github quirks that resolve comments and other branches that break while rebasing. These thing happen to us all the time and I would like to reduce the number of points of possible failures. We shouldn't rebase unless another branch contains a changeset that is needed in the current one we're working on. If master changes in the meanwhile when we work on the PR - it does not infer that a rebase is in need. If our CI runs a `merge` build and not a `push` build the incompatibilities should pop up and be shown on the CI build.

I agree that this point somehow is in conflict with point 1, but the idea is to give guidelines to contributors while allowing core members a bit more flexibility according to fluctuating needs. There's no silver bullet here but we can at least minimize error by new and external contributors.

@nolash you are welcome to PR or start a document on development best practices. I really don't wanna make a PR template into a gocode best practices knowledge-base

**Comment by zelig** [comment]
I would add:
- before setting the ready for review label, make sure all checks passed. To try the ci locally, use  `go run ./build/ci.go lint`, `go run ./build/ci.go test`.
- for merging the PR must be reviewed and approved by at least two others and approved by all parties that submitted a review earlier even if there are 2 other approvals
- you do not need to create an issue on a PR you submit (early or complete, cos you would not anyway) but if there exists one, reference it with the appropriate magic keywords, e.g., `fixes #1501`
- consider submitting changes in other packages that are dependecies of a PR in separate PRs
- assign to yourself or whoever you delegate
- make sure the work is shown on a board by choosing the project
- use labels   
   - review status: in progress, ready for review, ready for another review
   - it is prudent to use 'in progress' before the PR is ready, but not needed
   - it is imperative to use 'ready for review'
   - in subsequent iterations, be brave and put back in progress, and ready for review
- PRs are an excellent interface to discuss changes, so sometimes you open PRs that are not meant to be merged. Use the 'do not merge' label for PR that are  
      - experimental (e.g., https://github.com/ethersphere/swarm/pull/1676) or 
      - demonstrative (https://github.com/ethersphere/swarm/pull/427)
      - supplemental (alternative implementation, e.g,  https://github.com/ethersphere/swarm/pull/1680)
- put on the board and groom the board regularly, esp blocker

File: pull_request_template.md
Comment (zelig) [reply pull_request_template.md]: as @nolash says it is in order for a second reviewer to judge if submitter works on fixes atm.

File: pull_request_template.md
Comment (nolash) [reply pull_request_template.md]: @acud my point is not a go-specific one. Conflicts would happen in any kind of source. By keeping separate temporary files, rebase is even less likely. This is also aligned with your announced goal:

> My idea is to avoid force pushing as much as possible

**Comment by nolash** [comment]
> for merging the PR must be reviewed and approved by at least two others and approved by all parties that submitted a review earlier even if there are 2 other approvals

@zelig @acud we should also add a reminder that before merge smoke tests on cluster should have been run and passed.

**Comment by nolash** [comment]
@acud is it a conscious choice that you have omitted encouragement to keep the PR deltas small?

**Comment by nolash** [comment]
> 9b. If there is strong difference of opinion, please double check that you have understood all proposed arguments in the process of discussion. Unresolved differences are resolved by simple majority of author and the two reviewers. Matters of taste are at the discretion of the author.

How about this? @acud

**Comment by nolash** [comment]
> we should also add a reminder that before merge smoke tests on cluster should have been run and passed.

...though in case of wholly external contributors this is not possible, I guess.

**Comment by acud** [comment]
@zelig @nolash I have nothing against your propositions but may I remind you that these guidelines should appear in the following (tiny) textbox:
2019-09-04-145102_1920x1080_scrot
Not sure how much text you wanna have in there but I feel that if we are not concise with a readable checklist people will just not read what's in the box. We are better off with having this in some wiki page/documentation page.

**Pull request merged**
